cpp/src/arrow/compute/kernels/take.cc

```diff
@@ -99,5 +100,95 @@ Status Take(FunctionContext* ctx, const Datum& values, const Datum& indices,
   return kernel->Call(ctx, values, indices, out);
 }
 
+Status Take(FunctionContext* ctx, const ChunkedArray& values, const Array& indices,
+            const TakeOptions& options, std::shared_ptr<ChunkedArray>* out) {
+  auto num_chunks = values.num_chunks();
+  std::vector<std::shared_ptr<arrow::Array>> new_chunks(1); // Hard-coded 1 for now
+  std::shared_ptr<arrow::Array> current_chunk;
+
+  // Case 1: `values` has a single chunk, so just use it
+  if (num_chunks == 1) {
+    current_chunk = values.chunk(0);
+  } else {
+    // TODO Case 2: See if all `indices` fall in the same chunk and call Array Take on it
+    // See https://github.com/apache/arrow/blob/6f2c9041137001f7a9212f244b51bc004efc29af/r/src/compute.cpp#L123-L151
+    // TODO Case 3: If indices are sorted, can slice them and call Array Take
+
+    // Case 4: Else, concatenate chunks and call Array Take
+    RETURN_NOT_OK(Concatenate(values.chunks(), default_memory_pool(), &current_chunk));
+  }
+  // Call Array Take on our single chunk
+  RETURN_NOT_OK(Take(ctx, *current_chunk, indices, options, &new_chunks[0]));
+  *out = std::make_shared<arrow::ChunkedArray>(std::move(new_chunks));
+  return Status::OK();
+}
+
+Status Take(FunctionContext* ctx, const ChunkedArray& values, const ChunkedArray& indices,
+            const TakeOptions& options, std::shared_ptr<ChunkedArray>* out) {
+  auto num_chunks = indices.num_chunks();
+  std::vector<std::shared_ptr<arrow::Array>> new_chunks(num_chunks);
+  std::shared_ptr<arrow::ChunkedArray> current_chunk;
+
+  for (int i = 0; i < num_chunks; i++) {
+    // Take with that indices chunk
+    RETURN_NOT_OK(Take(ctx, values, *indices.chunk(i), options, &current_chunk));
+    // Concatenate the result to make a single array for this chunk
+    RETURN_NOT_OK(Concatenate(current_chunk->chunks(), default_memory_pool(), &new_chunks[i]));
```

Comment:
(of course it's a moot point until `Take(ChunkedArray values, Array indices)` can also avoid concatenation)

Reply:
The point of this concatenate was that the resulting chunks should correspond to the chunks defined by `indices`, as @jorisvandenbossche suggested. This, of course, gets us back to the original discussion of what chunks are for, whether they are purely an internal implementation detail or something that users should govern, what "optimal" chunking is, etc.

Reply:
I think optimal chunking for output from a kernel is whatever allows the consumer greatest control over allocation and other performance overhead. Based on this: concatenation should be kept to a minimum since that generates new allocations instead of cheaply slicing existing ones. As a secondary consideration, the chunked array should have as few chunks as possible since large contiguous chunks can be processed more efficiently than lots of short chunks. Based on that: an output chunked array should not contain empty chunks.